File: 792. Number of Matching Subsequences/Solution.java
class Solution {
    public int numMatchingSubseq(String s, String[] words) {
        List<List<String>> bucket = new LinkedList<>();
        for (int i = 0; i < 26; i++) {
            bucket.add(new LinkedList<>());
        }
        
        for (String w : words) {
            bucket.get(w.charAt(0) - 'a').add(w);
        }
        
        int n = s.length();
        int res = 0;
        for (int i = 0; i < n; i++) {
            List<String> list = bucket.get(s.charAt(i) - 'a');
            int size = list.size();
            for (int j = size - 1; j >= 0; j--) {
                String temp = list.get(j);
                if (temp.length() == 1) {
                    res++;
                } else {
                    temp = temp.substring(1, temp.length());
                    bucket.get(temp.charAt(0) - 'a').add(temp); 
                }
                list.remove(j);
            }
        }
        return res;
    }
}

File: 792. Number of Matching Subsequences/Solution.cpp
typedef pair<int, int> PII;

#define x first
#define y second

class Solution {
public:
    int numMatchingSubseq(string s, vector<string>& words) {
        vector<PII> ps[26];
        int res = 0;
        for (int i = 0; i < words.size(); i++)
        {
            ps[words[i][0] - 'a'].push_back({i, 0});
        }
        for (auto c : s)
        {
            vector<PII> buf;
            for (auto& p : ps[c - 'a'])
            {
                if (p.y == words[p.x].size() - 1) res++;
                else buf.push_back({p.x, p.y + 1});
            }
            ps[c - 'a'].clear();
            for (auto& b : buf)
            {
                ps[words[b.x][b.y] - 'a'].push_back(b);
            }
        }
        return res;
    }
};
